File: tests/positives/mcp_client/python/detect-mcp-client.py
# POSITIVE TEST CASES - Should be detected by the MCP client rule
# These are based on the official MCP Python SDK patterns

import asyncio
from pydantic import AnyUrl

# OFFICIAL MCP SDK IMPORTS - should match
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from mcp.client.session import ClientSession
from mcp.client.streamable_http import streamablehttp_client
from mcp.client.auth import OAuthClientProvider
from mcp.types import PromptReference, ResourceTemplateReference
from mcp import types
from mcp.shared.auth import OAuthToken, OAuthClientInformationFull

# MCP-SPECIFIC PARAMETER CLASSES - should match
server_params = StdioServerParameters(
    command="uv",
    args=["run", "server", "fastmcp_quickstart", "stdio"],
    env={"UV_INDEX": ""}
)

# Alternative parameter initialization
stdio_params = StdioServerParameters(
    command="python",
    args=["-m", "my_mcp_server"]
)

# OFFICIAL STDIO CLIENT USAGE - should match
async def stdio_client_example():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()
            tools = await session.list_tools()
            return tools

async def stdio_client_variant():
    # Direct usage without assignment
    async with stdio_client(StdioServerParameters(command="mcp-server", args=[])) as (read, write):
        session = ClientSession(read, write)
        await session.initialize()

# OFFICIAL STREAMABLE HTTP CLIENT USAGE - should match
async def streamable_http_example():
    async with streamablehttp_client("http://localhost:8000/mcp") as (read_stream, write_stream, _):
        async with ClientSession(read_stream, write_stream) as session:
            await session.initialize()
            tools = await session.list_tools()

async def streamable_http_variant():
    client_connection = streamablehttp_client("http://mcp-server:3000/mcp")

# CLIENT SESSION PATTERNS - should match
async def client_session_examples():
    async with stdio_client(server_params) as (read, write):
        # Basic ClientSession usage
        session = ClientSession(read, write)
        await session.initialize()

        # Session with context manager
        async with ClientSession(read, write) as session:
            await session.initialize()

# MCP-SPECIFIC TOOL OPERATIONS - should match
async def tool_operations():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            # List tools
            tools = await session.list_tools()
            available_tools = await session.list_tools()

            # Call tools
            result = await session.call_tool("weather", {"location": "Seattle"})
            weather_result = await session.call_tool("get_weather", {"city": "Tokyo"})

# MCP-SPECIFIC RESOURCE OPERATIONS - should match
async def resource_operations():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            # List resources
            resources = await session.list_resources()
            available_resources = await session.list_resources()

            # Read resources
            content = await session.read_resource(AnyUrl("file:///greeting.txt"))
            document = await session.read_resource(AnyUrl("documents://readme.md"))

            # List resource templates
            templates = await session.list_resource_templates()

# MCP-SPECIFIC PROMPT OPERATIONS - should match
async def prompt_operations():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            # List prompts
            prompts = await session.list_prompts()
            available_prompts = await session.list_prompts()

            # Get prompts
            prompt = await session.get_prompt("greet_user", arguments={"name": "Alice", "style": "friendly"})
            greeting = await session.get_prompt("generate_greeting", {"user": "Bob"})

# MCP COMPLETION API - should match (very specific to MCP)
async def completion_operations():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            # Resource template completion
            result = await session.complete(
                ref=ResourceTemplateReference(type="ref/resource", uri="github://owner/{repo}"),
                argument={"name": "repo", "value": ""}
            )

            # Prompt completion
            completion = await session.complete(
                ref=PromptReference(type="ref/prompt", name="greet_user"),
                argument={"name": "style", "value": ""}
            )

# MCP NOTIFICATION SENDING - should match
async def notification_operations():
    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            await session.initialize()

            # Send resource notifications
            await session.send_resource_updated(AnyUrl("file:///config.json"))
            await session.send_resource_list_changed()

# MCP TYPES USAGE - should match
def mcp_types_usage():
    # Type imports and usage
    prompt_ref = PromptReference(type="ref/prompt", name="my_prompt")
    resource_ref = ResourceTemplateReference(type="ref/resource", uri="docs://{file}")
    file_url = AnyUrl("file:///documents/readme.md")

# OAUTH PATTERNS SPECIFIC TO MCP - should match
async def oauth_patterns():
    oauth_provider = OAuthClientProvider(
        server_url="http://localhost:8001",
        client_metadata=types.OAuthClientMetadata(
            client_name="Example MCP Client",
            redirect_uris=[AnyUrl("http://localhost:3000/callback")],
            grant_types=["authorization_code", "refresh_token"],
            response_types=["code"],
            scope="user"
        )
    )

# REAL-WORLD COMPREHENSIVE EXAMPLE - should match
async def comprehensive_mcp_client():
    """Complete MCP client implementation example"""
    server_params = StdioServerParameters(
        command="uv",
        args=["run", "my-mcp-server"],
        env={"MCP_ENV": "production"}
    )

    async with stdio_client(server_params) as (read, write):
        async with ClientSession(read, write) as session:
            # Initialize connection
            await session.initialize()

            # Discover capabilities
            tools = await session.list_tools()
            resources = await session.list_resources()
            prompts = await session.list_prompts()
            templates = await session.list_resource_templates()

            # Execute tool
            if tools.tools:
                result = await session.call_tool(
                    tools.tools[0].name,
                    {"input": "test data"}
                )

            # Read resource
            if resources.resources:
                content = await session.read_resource(resources.resources[0].uri)

            # Get prompt
            if prompts.prompts:
                prompt_result = await session.get_prompt(
                    prompts.prompts[0].name,
                    {"context": "example"}
                )

            # Completion example
            if templates.resourceTemplates:
                completion = await session.complete(
                    ref=ResourceTemplateReference(
                        type="ref/resource",
                        uri=templates.resourceTemplates[0].uriTemplate
                    ),
                    argument={"name": "param", "value": "partial"}
                )

# STREAMABLE HTTP WITH AUTH - should match
async def streamable_http_with_auth():
    oauth_auth = OAuthClientProvider(
        server_url="http://localhost:8001",
        client_metadata=types.OAuthClientMetadata(
            client_name="Auth MCP Client",
            redirect_uris=[AnyUrl("http://localhost:3000/callback")]
        )
    )

    async with streamablehttp_client("http://localhost:8001/mcp", auth=oauth_auth) as (read, write, _):
        async with ClientSession(read, write) as session:
            await session.initialize()
            tools = await session.list_tools()

if __name__ == "__main__":
    # Run examples
    asyncio.run(stdio_client_example())
    asyncio.run(comprehensive_mcp_client())

File: tests/positives/mcp_client/typescript/detect-mcp-client.ts
// Positive test cases for TypeScript MCP client detection

// Test case 1: Core MCP client imports
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { Client as MCPClient } from "@modelcontextprotocol/sdk/client";

// Test case 2: MCP transport imports
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { StdioClientTransport as StdioTransport } from "@modelcontextprotocol/sdk/client/stdio";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { SSEClientTransport as SSETransport } from "@modelcontextprotocol/sdk/client/sse";
import { HttpClientTransport } from "@modelcontextprotocol/sdk/client/http.js";
import { HttpClientTransport as HttpTransport } from "@modelcontextprotocol/sdk/client/http";

// Test case 3: MCP types and schema imports
import {
  CallToolResultSchema,
  ListToolsResultSchema,
  ListResourcesResultSchema,
  GetPromptResultSchema
} from "@modelcontextprotocol/sdk/types.js";

import {
  CallToolResultSchema as CallToolSchema,
  ListToolsResultSchema as ListToolsSchema
} from "@modelcontextprotocol/sdk/types";

// Test case 4: Multi-server management import
import { MultiServerManager } from "@modelcontextprotocol/sdk/client";

// Test case 5: Basic MCP client instantiation
function createBasicMCPClient() {
  const client = new Client({
    name: "mcp-client",
    version: "1.0.0"
  });

  return client;
}

// Test case 6: MCP client with capabilities
function createMCPClientWithCapabilities() {
  const client = new Client(
    { name: "example-client", version: "1.0.0" },
    {
      capabilities: {
        prompts: {},
        resources: {},
        tools: {}
      }
    }
  );

  return client;
}

// Test case 7: Class-based MCP client implementation
class MCPClientManager {
  private mcp: Client;
  private transport: StdioClientTransport | SSEClientTransport;

  constructor() {
    this.mcp = new Client({
      name: "mcp-client-cli",
      version: "1.0.0"
    });
  }

  // Test case 8: STDIO transport setup
  async connectToStdioServer(serverScript: string): Promise<void> {
    this.transport = new StdioClientTransport({
      command: "python",
      args: [serverScript]
    });

    await this.mcp.connect(this.transport);
  }

  // Test case 9: SSE transport setup
  async connectToSSEServer(url: string): Promise<void> {
    const transport = new SSEClientTransport(url);
    await this.mcp.connect(transport);
  }

  // Test case 10: HTTP transport setup
  async connectToHttpServer(baseUrl: string): Promise<void> {
    const transport = new HttpClientTransport({
      url: baseUrl
    });
    await this.mcp.connect(transport);
  }

  // Test case 11: MCP tool operations
  async performToolOperations(): Promise<void> {
    // List available tools
    const tools = await this.mcp.listTools();
    const availableTools = await this.mcp.listTools();

    // Call specific tools
    const result = await this.mcp.callTool("calculate", { a: 5, b: 3 });
    const weatherResult = await this.mcp.callTool("get_weather", {
      location: "San Francisco"
    });
  }

  // Test case 12: MCP resource operations
  async performResourceOperations(): Promise<void> {
    // List available resources
    const resources = await this.mcp.listResources();
    const availableResources = await this.mcp.listResources();

    // Read specific resources
    const content = await this.mcp.readResource("file:///example.txt");
    const document = await this.mcp.readResource({
      uri: "file:///document.pdf"
    });
  }

  // Test case 13: MCP prompt operations
  async performPromptOperations(): Promise<void> {
    // List available prompts
    const prompts = await this.mcp.listPrompts();
    const availablePrompts = await this.mcp.listPrompts();

    // Get specific prompts
    const prompt = await this.mcp.getPrompt("code_review", {
      language: "typescript"
    });
    const greeting = await this.mcp.getPrompt("greeting", {});
  }

  // Test case 14: Connection management
  async cleanup(): Promise<void> {
    if (this.mcp) {
      await this.mcp.close();
    }
  }
}

// Test case 15: Low-level MCP request patterns
async function performLowLevelRequests(client: Client): Promise<void> {
  // Tools requests
  const toolsResult = await client.request(
    { method: "tools/list" },
    ListToolsResultSchema
  );

  const callResult = await client.request(
    { method: "tools/call" },
    CallToolResultSchema
  );

  // Resources requests
  const resourcesResult = await client.request(
    { method: "resources/list" },
    ListResourcesResultSchema
  );

  const readResult = await client.request(
    { method: "resources/read" },
    ListResourcesResultSchema // Schema placeholder
  );

  // Prompts requests
  const promptsResult = await client.request(
    { method: "prompts/list" },
    GetPromptResultSchema
  );

  const getPromptResult = await client.request(
    { method: "prompts/get" },
    GetPromptResultSchema
  );
}

// Test case 16: Alternative transport configurations
function createAlternativeTransports() {
  // Node.js server
  const nodeTransport = new StdioClientTransport({
    command: "node",
    args: ["server.js"]
  });

  // NPX-based server
  const npxTransport = new StdioClientTransport({
    command: "npx",
    args: ["-y", "kubernetes-mcp-server@latest"]
  });

  // SSE with URL object
  const sseTransport = new SSEClientTransport(
    new URL("http://localhost:3000/sse")
  );

  return { nodeTransport, npxTransport, sseTransport };
}

// Test case 17: Multi-server management
class MultiServerMCPClient {
  private manager: MultiServerManager;

  constructor() {
    this.manager = new MultiServerManager();
  }

  async setupMultipleServers(): Promise<void> {
    // Setup would involve multiple server connections
    // This pattern is specific to MCP multi-server scenarios
  }
}

// Test case 18: MCP with AI integration pattern
import { Anthropic } from "@anthropic-ai/sdk";

class MCPAnthropicClient {
  private mcp: Client;
  private anthropic: Anthropic;

  constructor() {
    this.anthropic = new Anthropic();
    this.mcp = new Client({
      name: "mcp-anthropic-client",
      version: "1.0.0"
    });
  }

  async initializeMCPConnection(serverPath: string): Promise<void> {
    const transport = new StdioClientTransport({
      command: "python",
      args: [serverPath]
    });

    await this.mcp.connect(transport);
  }

  async callMCPTool(toolName: string, args: any): Promise<any> {
    const result = await this.mcp.callTool(toolName, args);
    return result;
  }
}

// Test case 19: Error handling with MCP operations
async function mcpWithErrorHandling(): Promise<void> {
  const client = new Client({
    name: "error-handling-client",
    version: "1.0.0"
  });

  const transport = new StdioClientTransport({
    command: "python",
    args: ["server.py"]
  });

  try {
    await client.connect(transport);
    const tools = await client.listTools();
    const result = await client.callTool("example_tool", {});
  } catch (error) {
    console.error("MCP operation failed:", error);
  } finally {
    await client.close();
  }
}

// Test case 20: Complex capabilities configuration
function createClientWithDetailedCapabilities() {
  const client = new Client(
    { name: "advanced-client", version: "2.0.0" },
    {
      capabilities: {
        prompts: {
          listChanged: true
        },
        resources: {
          subscribe: true,
          listChanged: true
        },
        tools: {
          listChanged: true
        }
      }
    }
  );

  return client;
}

// Export for potential use
export {
  MCPClientManager,
  MultiServerMCPClient,
  MCPAnthropicClient,
  createBasicMCPClient,
  createMCPClientWithCapabilities
};